File: dotNetTips.Utility.Standard.bak2/IO/DriveHelper.cs
// ***********************************************************************
// Assembly         : dotNetTips.Utility.Standard
// Author           : David McCarter
// Created          : 01-22-2017
//
// Last Modified By : David McCarter
// Last Modified On : 01-21-2017
// ***********************************************************************
// <copyright file="DriveHelper.cs" company="dotNetTips.Utility.Standard">
//     Copyright (c) dotNetTips.com - McCarter Consulting. All rights reserved.
// </copyright>
// <summary></summary>
// ***********************************************************************
/// <summary>
/// The IO namespace.
/// </summary>
namespace dotNetTips.Utility.Standard.IO
{
    /// <summary>
    /// Class DriveHelper.
    /// </summary>
    public static class DriveHelper
    {
        /// <summary>
        /// Finds the serial number of a drive.
        /// </summary>
        /// <param name="drive">Drive name.</param>
        /// <returns>Serial number.</returns>
        /// <remarks>This call could take some time. Recommend multi-threading.</remarks>
        //public string GetDriveSerialNumber(string drive)
        //{
        //    System.Diagnostics.Contracts.Contract.Requires<ArgumentNullException>(string.IsNullOrWhiteSpace(drive) == false);

        //    string driveSerial = string.Empty;

        //    //No matter what is sent in, get just the drive letter
        //    string driveFixed = System.IO.Path.GetPathRoot(drive).Replace("\\", string.Empty);
           
        //    //Perform Query
        //    using (var querySearch = new ManagementObjectSearcher(string.Format("SELECT VolumeSerialNumber FROM Win32_LogicalDisk Where Name = '{0}'", driveFixed)))
        //    {
        //        using (var queryCollection = querySearch.Get())
        //        {
        //            ManagementObject moItem;
        //            foreach (moItem in queryCollection)
        //            {
        //                driveSerial = (string)moItem.Item("VolumeSerialNumber");
        //                break; // TODO: might not be correct. Was : Exit For
        //            }
        //        }
        //    }

        //    return driveSerial;

        //}
    }
}

File: dotNetTips.Utility/IO/DriveHelper.vb
' ***********************************************************************
' Assembly         : dotNetTips
' Author           : David McCarter
' Created          : 12-10-2008
'
' Last Modified By : David McCarter
' Last Modified On : 05-11-2017
' ***********************************************************************
' <copyright file="DriveHelper.vb" company="McCarter Consulting - David McCarter">
'     David McCarter - dotNetTips.com © 2017
' </copyright>
' <summary></summary>
' *************************************************************************
Imports System.Management
Imports dotNetTips.Utility.Portable.OOP

Namespace IO
    ''' <summary>
    ''' Drive helper methods.
    ''' </summary>
    Public Module DriveHelper
        ''' <summary>
        ''' Finds the serial number of a drive.
        ''' </summary>
        ''' <param name="drive">Drive name.</param>
        ''' <returns>Serial number.</returns>
        ''' <remarks>This call could take some time. Recommend multi-threading.</remarks>
        Public Function GetDriveSerialNumber(ByVal drive As String) As String
            Encapsulation.TryValidateParam(drive, NameOf(drive))

            Dim driveSerial = String.Empty

            'No matter what is sent in, get just the drive letter
            Dim driveFixed = System.IO.Path.GetPathRoot(drive)
            driveFixed = Replace(driveFixed, "\", String.Empty)

            'Perform Query
            Using querySearch As New ManagementObjectSearcher(String.Format("SELECT VolumeSerialNumber FROM Win32_LogicalDisk Where Name = '{0}'", driveFixed))
                Using queryCollection As ManagementObjectCollection = querySearch.Get()

                    For Each moItem In queryCollection.AsParallel()
                        driveSerial = CStr(moItem.Item("VolumeSerialNumber"))
                        Exit For
                    Next

                End Using
            End Using

            Return driveSerial

        End Function

    End Module
End Namespace